eventlet/wsgi.py

```diff
@@ -607,11 +618,22 @@ def cap(x):
         finally:
             if hasattr(result, 'close'):
                 result.close()
-            request_input = self.environ['eventlet.input']
+            if request_input.should_send_hundred_continue:
+                # We just sent the final response, no 100 Continue. Client may or
+                # may not have started to send a body, and if we keep the connection
+                # open we've seen clients either
+                #   * send a body, then start a new request
+                #   * skip the body and go straight to a new request
+                # Looks like the most broadly compatible option is to close the
+                # connection and let the client retry.
+                # https://curl.se/mail/lib-2004-08/0002.html
+                # Note that we likely *won't* send a Connection: close header at this point
+                self.close_connection = 1
```

Comment:
this is really the crux of it; we've handled this response - if we didn't send a 100-continue and the client was expecting one that's it for this connection

does it matter if it was a zero byte POST/PUT? can you even set expect-100 in that case?

so up above where we're being cute with hasattr `__len__` and connection close headers; do we really want to close the connection, or do we want to set "should_send_hundred_continue" to False and let it discard?

Reply:
> does it matter if it was a zero byte POST/PUT? can you even set expect-100 in that case?

Expect-100 with zero body is valid, with especially clear semantic for PUT (rewrite with empty). It's not economic to close connection in zero body request case, but probably it's also very rare.

> so up above where we're being cute with hasattr __len__ and connection close headers; do we really want to close the connection, or do we want to set "should_send_hundred_continue" to False and let it discard?

My vote for close as more robust option.